src/Microsoft.DotNet.VersionTools/Dependencies/MSBuildPackageReferenceUpdater.cs

```diff
@@ -0,0 +1,187 @@
+        }
+
+        private string ReplaceAllDependencyVersions(
+            string input,
+            string projectJsonFile,
+            IEnumerable<DependencyBuildInfo> buildInfos,
+            out IEnumerable<DependencyChange> dependencyChanges)
+        {
+            XDocument documentRoot = XDocument.Parse(input);
+
+            dependencyChanges = FindAllDependencyProperties(documentRoot)
+                    .Select(dependencyProperty => ReplaceDependencyVersion(projectJsonFile, dependencyProperty, buildInfos))
+                    .Where(buildInfo => buildInfo != null)
+                    .ToArray();
+
+            StringWriter sw = new StringWriter();
+            documentRoot.Save(sw, SaveOptions.None);
+            return sw.ToString();
+        }
+
+        /// <summary>
+        /// Replaces the single dependency with the updated version, if it matches any of the
+        /// dependencies that need to be updated. Stops on the first updated value found.
+        /// </summary>
+        /// <returns>The BuildInfo used to change the value, or null if there was no change.</returns>
+        private DependencyChange ReplaceDependencyVersion(
+            string projectJsonFile,
+            XElement packageRefElement,
+            IEnumerable<DependencyBuildInfo> parsedBuildInfos)
+        {
+            string id = packageRefElement.Attribute("Include").Value;
+            Debug.Assert(id != null);
+            foreach (DependencyBuildInfo info in parsedBuildInfos)
+            {
+                foreach (PackageIdentity packageInfo in info.Packages)
+                {
+                    if (id != packageInfo.Id)
+                    {
+                        continue;
+                    }
+
+                    string oldVersion;
+                    // Support the new  "inline" Version attribute, or the old-style nested Version element.
+                    XAttribute versionAttribute = null;
+                    XElement versionElement = null;
+
+                    versionAttribute = packageRefElement.Attribute(XName.Get("Version", packageRefElement.GetDefaultNamespace().NamespaceName));
+                    if (versionAttribute == null)
+                    {
+                        versionElement = packageRefElement.Element(XName.Get("Version", packageRefElement.GetDefaultNamespace().NamespaceName));
+                    }
+
+                    if (versionAttribute != null)
+                    {
+                        oldVersion = versionAttribute.Value;
+                    }
+                    else
+                    {
+                        if (versionElement == null)
+                        {
+                            Trace.TraceWarning($"Couldn't parse the version information for package '{id}' in '{projectJsonFile}'. Skipping.");
+                            continue;
+                        }
+
+                        oldVersion = versionElement.Value;
+                    }
```

Comment:
This structure isn't immediately clear--`if (a!=null) else if (e!=null) else fail` seems better to me.